src/client/api/metrics.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * (C) Copyright 2020-2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/utsname.h>
+#include <daos/common.h>
+#include <daos/job.h>
+#include <daos/tls.h>
+#include <daos/metrics.h>
+#include <gurt/telemetry_common.h>
+#include <gurt/telemetry_consumer.h>
+#include <gurt/telemetry_producer.h>
+
+#define INIT_JOB_NUM		1024
+bool daos_client_metric;
+bool daos_client_metric_retain;
+
+#define MAX_IDS_SIZE(num) (num * D_TM_METRIC_SIZE)
+/* The client side metrics structure looks like
+ * root/job_id/pid/....
+ */
+int
+dc_tm_init(void)
+{
+	int    metrics_tag;
+	pid_t  pid;
+	int    rc;
+
+	d_getenv_bool(DAOS_CLIENT_METRICS_ENABLE, &daos_client_metric);
+	if (!daos_client_metric)
+		return 0;
+
+	rc = dc_tls_key_create();
+	if (rc)
+		D_GOTO(out, rc);
+
+	metrics_tag = D_TM_OPEN_OR_CREATE;
+	d_getenv_bool(DAOS_CLIENT_METRICS_RETAIN, &daos_client_metric_retain);
+	if (daos_client_metric_retain)
+		metrics_tag |= D_TM_RETAIN_SHMEM;
+	else
+		metrics_tag |= D_TM_RETAIN_SHMEM_IF_NON_EMPTY;
+
+	rc = d_tm_init(DC_TM_JOB_ROOT_ID, MAX_IDS_SIZE(INIT_JOB_NUM), metrics_tag);
+	if (rc != 0) {
+		DL_ERROR(rc, "init job root id.");
+		return rc;
+	}
+
+	D_INFO("INIT %s metrics\n", dc_jobid);
+	rc = d_tm_add_ephemeral_dir(NULL, MAX_IDS_SIZE(INIT_JOB_NUM), "%s", dc_jobid);
+	if (rc != 0 && rc != -DER_EXIST) {
+		DL_ERROR(rc, "add metric %s failed.\n", dc_jobid);
```

Comment:
You need to resolve these.  DL_ERROR will add the formatting for you so remove the ".\n" at the end.
```suggestion
		DL_ERROR(rc, "add metric %s failed", dc_jobid);
```

Reply:
> You need to resolve these. DL_ERROR will add the formatting for you so remove the ".\n" at the end.

sure, thanks